packages/eslint-plugin/src/rules/strict-enums.ts

```diff
@@ -0,0 +1,806 @@
+          return true;
+        }
+      }
+
+      /**
+       * Allow function calls that match one of the types in a union, like the
+       * following:
+       *
+       * ```ts
+       * function useApple(fruitOrNull: Fruit | null) {}
+       * useApple(null);
+       * ```
+       */
+      const argumentSubTypesSet = new Set(argumentSubTypes);
+      const parameterSubTypesSet = new Set(parameterSubTypes);
+      if (setHasAnyElementFromSet(argumentSubTypesSet, parameterSubTypesSet)) {
+        return false;
+      }
+
+      /**
+       * Allow function calls that have a base enum that match the function
+       * type, like the following:
+       *
+       * ```ts
+       * function useFruit(fruit: Fruit) {}
+       * useFruit(Fruit.Apple);
+       * ```
+       */
+      if (setHasAnyElementFromSet(argumentEnumTypes, parameterEnumTypes)) {
+        return false;
+      }
+
+      return true;
+    }
+
+    // ------------------
+    // AST node callbacks
+    // ------------------
+
+    return {
+      /** When something is assigned to a variable. */
+      AssignmentExpression(node): void {
+        const leftType = getTypeFromNode(node.left);
+        const rightType = getTypeFromNode(node.right);
+
+        if (isAssigningNonEnumValueToEnumVariable(leftType, rightType)) {
+          context.report({
+            node,
+            messageId: 'mismatchedAssignment',
+          });
+        }
+      },
+
+      /** When a comparison between two things happen. */
+      BinaryExpression(node): void {
+        const leftType = getTypeFromNode(node.left);
+        const rightType = getTypeFromNode(node.right);
+
+        if (isMismatchedEnumComparison(node.operator, leftType, rightType)) {
+          context.report({
+            node,
+            messageId: 'mismatchedComparison',
+          });
+        }
+      },
+
+      /** When a function is invoked. */
+      CallExpression(node): void {
+        checkCallExpression(node);
+      },
+
+      /** When something is instantiated with the "new" keyword. */
+      NewExpression(node): void {
+        /**
+         * We need to perform the exact same checks on a class constructor
+         * invocation as we do on a normal function invocation.
+         */
+        checkCallExpression(node);
+      },
+
+      /** When a unary operator is invoked. */
+      UpdateExpression(node): void {
+        const argumentType = getTypeFromNode(node.argument);
+
+        /**
+         * Disallow using enums with unary operators, like the following:
+         *
+         * ```ts
+         * const fruit = Fruit.Apple;
+         * fruit++;
+         * ```
+         */
+        if (hasEnumTypes(argumentType)) {
+          context.report({
+            node,
+            messageId: 'incorrectIncrement',
+          });
+        }
+      },
+
+      /** When a new variable is created. */
+      VariableDeclaration(node): void {
+        for (const declaration of node.declarations) {
```

Comment:
nitpick - we don't care about the declaration itself, just the declarators.

```suggestion
      VariableDeclarator(node): void {
```

This is good because it saves you manually iterating and it means your report will be cleaner.

I.e. instead of:
```
const x = 1,
^^^^^^^^^^^^
      y: Foo = 99;
^^^^^^^^^^^^^^^^^
```
(where `^` is the report underline)

you'll instead have:
```
const x = 1,
      y: Foo = 99;
      ^^^^^^^^^^^
```

Which is clearer and more accurate!